src/physics/em/detail/MuBremsstrahlungInteractor.i.hh

```diff
@@ -0,0 +1,198 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file MuBremsstrahlungInteractor.i.hh
+//---------------------------------------------------------------------------//
+
+#include "base/ArrayUtils.hh"
+#include "base/Constants.hh"
+#include "base/Range.hh"
+#include "random/distributions/UniformRealDistribution.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION MuBremsstrahlungInteractor::MuBremsstrahlungInteractor(
+    const MuBremsstrahlungInteractorPointers& shared,
+    const ParticleTrackView&                  particle,
+    const Real3&                              inc_direction,
+    StackAllocator<Secondary>&                allocate,
+    const MaterialView&                       material,
+    ElementComponentId                        elcomp_id)
+    : shared_(shared)
+    , inc_energy_(particle.energy().value())
+    , inc_direction_(inc_direction)
+    , allocate_(allocate)
+    , element_(material.element_view(elcomp_id))
+    , inc_mass_(particle.mass().value())
+{
+    CELER_EXPECT(inc_energy_ >= shared_.min_incident_energy()
+                 && inc_energy_ <= shared_.max_incident_energy());
+    CELER_EXPECT(particle.particle_id() == shared_.mu_minus_id 
+                 || particle.particle_id() == shared_.mu_plus_id); 
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample using the Muon Bremsstrahlung model.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MuBremsstrahlungInteractor::operator()(Engine& rng)
+{
+    // Allocate space for gamma
+    Secondary* secondaries = this->allocate_(1); 
+    if (secondaries == nullptr)
+    {
+        // Failed to allocate space for a secondary
+        return Interaction::from_failure();
+    }
+
+    const real_type min_inc_kinetic_energy 
+            = min(inc_energy_.value(), shared_.min_incident_energy().value());
+    const real_type func_1 = min_inc_kinetic_energy 
+                * this->differential_cross_section(min_inc_kinetic_energy);
+    const real_type x_min = std::log(min_inc_kinetic_energy);
+    const real_type x_max = std::log(inc_energy_.value() 
+                                        / min_inc_kinetic_energy);
+                                        
+    real_type ln_epsilon, epsilon, func_2;
+    UniformRealDistribution<real_type> p;
+
+    do
+    {
+        ln_epsilon = x_min + p(rng) * x_max;
+        epsilon    = std::exp(ln_epsilon);
+        func_2     = epsilon * this->differential_cross_section(epsilon);
+    } while(func_2 < func_1 * p(rng));
```

Comment:
I think we could simplify this rejection loop slightly since `ln_epsilon` is only used to calculate `epsilon` inside the loop and `func_2` is only used in the conditional:
```
    real_type epsilon;
    do  
    {   
        epsilon = std::exp(x_min + generate_canonical(rng) * x_max);
    } while (BernoulliDistribution(
        1 - epsilon * this->differential_cross_section(epsilon) / func_1)(rng));
```

Reply:
You should be able to simplify it even further since the sampling of epsilon is a sample of the pdf `f(x) = 1/x`:
```c++
ReciprocalDistribution sample_epsilon(min_inc_kinetic_energy , inc_energy_.value());

real_type epsilon;
do  
{ 
    epsilon = sample_epsilon(rng);
} while (...);
```

Reply:
Also `BernoulliDistribution(1 - x)(rng)` has the same expected value as `!BernoulliDistribution(x)(rng)` so that'll also help :)

I really should add another `RejectionSampler(x, xmax)(rng)` that's equivalent to `!Bernoulli(x / xmax)`...

Reply:
Ahh good call! 🙂